// d2-ui/lib/auto-complete/AutoComplete.component.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _textField = require('material-ui/lib/text-field');

var _textField2 = _interopRequireDefault(_textField);

var _Action = require('../action/Action');

var _Action2 = _interopRequireDefault(_Action);

var _rx = require('rx');

var _rx2 = _interopRequireDefault(_rx);

var _d = require('d2/lib/d2');

var _d2 = _interopRequireDefault(_d);

var _list = require('material-ui/lib/lists/list');

var _list2 = _interopRequireDefault(_list);

var _listItem = require('material-ui/lib/lists/list-item');

var _listItem2 = _interopRequireDefault(_listItem);

var _paper = require('material-ui/lib/paper');

var _paper2 = _interopRequireDefault(_paper);

var _Translate = require('../i18n/Translate.mixin');

var _Translate2 = _interopRequireDefault(_Translate);

var _loglevel = require('loglevel');

var _loglevel2 = _interopRequireDefault(_loglevel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

_d.config.i18n.strings.add('members');
_d.config.i18n.strings.add('search_for_user_groups');

function searchByForModel(searchBy, modelTypeToSearch, valueToSearchFor) {
    var options = arguments.length <= 3 || arguments[3] === undefined ? {} : arguments[3];

    if (!Boolean(modelTypeToSearch) || !Boolean(valueToSearchFor)) {
        _loglevel2.default.warn('forType property and value should be provided to be able to show results');

        return _rx.Observable.just([]);
    }

    var searchQueryRequest = _d2.default.getInstance().then(function (d2) {
        return d2.models[modelTypeToSearch];
    }).then(function (modelType) {
        return modelType.filter().on(searchBy).ilike(valueToSearchFor);
    }).then(function (modelTypeWithFilter) {
        return modelTypeWithFilter.list(options);
    }).then(function (collection) {
        return collection.toArray();
    }).catch(function (error) {
        return _loglevel2.default.error(error);
    });

    return _rx.Observable.fromPromise(searchQueryRequest);
}

exports.default = (0, _react.createClass)({
    propTypes: {
        actions: _react.PropTypes.object,
        forType: _react.PropTypes.string.isRequired,
        onSuggestionClicked: _react.PropTypes.func.isRequired,
        closeOnItemClicked: _react.PropTypes.bool,
        clearValueOnItemClicked: _react.PropTypes.bool,
        filterForSuggestions: _react.PropTypes.func
    },

    mixins: [_Translate2.default],

    getDefaultProps: function getDefaultProps() {
        return {
            actions: _Action2.default.createActionsFromNames(['loadAutoCompleteSuggestions']),
            debounceTime: 500,
            propertyToSearchBy: 'displayName',
            scheduler: _rx.Scheduler.default,
            closeOnItemClicked: true,
            clearValueOnItemClicked: true
        };
    },

    //
    getInitialState: function getInitialState() {
        return {
            showAutoComplete: false,
            autoCompleteValues: [],
            loadingSuggestions: false
        };
    },
    componentWillMount: function componentWillMount() {
        var _this = this;

        var _props = this.props;
        var actions = _props.actions;
        var forType = _props.forType;


        this.disposable = actions.loadAutoCompleteSuggestions.map(function (_ref) {
            var event = _ref.data;
            return event.target.value;
        }).tap(function (value) {
            return _this.setState({
                loadingSuggestions: true,
                showAutoComplete: Boolean(value)
            });
        }).debounce(this.props.debounceTime, this.props.scheduler).distinctUntilChanged()
        // TODO: Do not hardcore these fields to search for
        .map(function (valueToSearchFor) {
            return searchByForModel(_this.props.propertyToSearchBy, forType, valueToSearchFor, { fields: 'id,displayName|rename(name),users::size', pageSize: 10 });
        }).concatAll().map(function (suggestions) {
            return Array.isArray(suggestions) ? suggestions.filter(_this.props.filterForSuggestions || _rx.helpers.identity) : [];
        }).map(function (suggestions) {
            return suggestions.slice(0, 5);
        }).subscribe(function (autoCompleteValues) {
            return _this.setState({
                autoCompleteValues: autoCompleteValues,
                loadingSuggestions: false
            });
        }, function (errorMessage) {
            return _loglevel2.default.error(errorMessage);
        });
    },
    componentWillUnmount: function componentWillUnmount() {
        this.disposable && this.disposable.dispose();
    },
    onSuggestionClick: function onSuggestionClick(item) {
        var _this2 = this;

        return function (event) {
            var _props2 = _this2.props;
            var closeOnItemClicked = _props2.closeOnItemClicked;
            var clearValueOnItemClicked = _props2.clearValueOnItemClicked;
            var onSuggestionClicked = _props2.onSuggestionClicked;


            if (closeOnItemClicked) {
                _this2.refs.autoCompleteField.focus();
            }

            if (clearValueOnItemClicked) {
                _this2.refs.autoCompleteField.setValue('');
                _this2.props.actions.loadAutoCompleteSuggestions({
                    target: { value: '' }
                });
            }

            _this2.setState({
                showAutoComplete: !closeOnItemClicked
            });

            if (onSuggestionClicked) {
                onSuggestionClicked(item, event);
            }
        };
    },


    // TODO: Allow the component user to specify how to render the list items or at least the primary and secondary texts
    renderAutoCompleteSuggestions: function renderAutoCompleteSuggestions() {
        var _this3 = this;

        return _react2.default.createElement(
            'div',
            { style: { position: 'absolute', zIndex: 100 } },
            _react2.default.createElement(
                _paper2.default,
                null,
                _react2.default.createElement(
                    _list2.default,
                    null,
                    this.state.autoCompleteValues.map(function (userGroup) {
                        return _react2.default.createElement(_listItem2.default, {
                            primaryText: userGroup.name,
                            secondaryText: userGroup.users + ' ' + _this3.getTranslation('members'),
                            innerDivStyle: { paddingTop: '.5rem', paddingBottom: '.5rem' },
                            onClick: _this3.onSuggestionClick(userGroup)
                        });
                    })
                )
            )
        );
    },
    render: function render() {
        var _props3 = this.props;
        var actions = _props3.actions;
        var forType = _props3.forType;

        var other = _objectWithoutProperties(_props3, ['actions', 'forType']);

        return _react2.default.createElement(
            'div',
            { style: { position: 'relative' }, onClick: function onClick(event) {
                    return event.stopPropagation();
                } },
            _react2.default.createElement(_textField2.default, _extends({
                ref: 'autoCompleteField' }, other, {
                onChange: actions.loadAutoCompleteSuggestions,
                hintText: this.getTranslation('search_for_user_groups'),
                fullWidth: true
            })),
            this.state.showAutoComplete && !this.state.loadingSuggestions ? this.renderAutoCompleteSuggestions() : null
        );
    }
});
