// d2-ui/lib/app-header/menus/AppMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _HeaderMenu = require('./HeaderMenu');

var _HeaderMenu2 = _interopRequireDefault(_HeaderMenu);

var _HeaderMenuItem = require('./HeaderMenuItem');

var _HeaderMenuItem2 = _interopRequireDefault(_HeaderMenuItem);

var _flatButton = require('material-ui/lib/flat-button');

var _flatButton2 = _interopRequireDefault(_flatButton);

var _apps = require('material-ui/lib/svg-icons/navigation/apps');

var _apps2 = _interopRequireDefault(_apps);

var _addD2Context = require('../../component-helpers/addD2Context');

var _addD2Context2 = _interopRequireDefault(_addD2Context);

var _colors = require('material-ui/lib/styles/colors');

var _d = require('d2/lib/d2');

var _getBaseUrlFromD2ApiUrl = require('../getBaseUrlFromD2ApiUrl');

var _getBaseUrlFromD2ApiUrl2 = _interopRequireDefault(_getBaseUrlFromD2ApiUrl);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// App menu strings to be translated
_d.config.i18n.strings.add('more_applications');

var styles = {
    moreAppsButton: {
        marginTop: '1rem',
        width: '375px',
        display: 'block',
        textAlign: 'center'
    }
};

function getBaseUrl(d2) {
    if (d2.Api) {
        return (0, _getBaseUrlFromD2ApiUrl2.default)(d2);
    }
    return './'; // Get old base url from local storage
}

var AppMenu = (0, _addD2Context2.default)(function AppMenu(props, _ref) {
    var d2 = _ref.d2;

    var menuItems = props.items.map(function (item, index) {
        return _react2.default.createElement(_HeaderMenuItem2.default, _extends({ key: index }, item));
    });
    var moreAppsButton = _react2.default.createElement(
        _flatButton2.default,
        { style: styles.moreAppsButton, linkButton: true, href: getBaseUrl(d2) + '/dhis-web-commons-about/modules.action' },
        d2.i18n.getTranslation('more_applications')
    );

    return _react2.default.createElement(
        _HeaderMenu2.default,
        {
            name: _react2.default.createElement(
                'div',
                null,
                _react2.default.createElement(_apps2.default, { color: _colors.white })
            ),
            moreButton: moreAppsButton,
            rowItemCount: props.rowItemCount,
            columnItemCount: props.columnItemCount
        },
        menuItems
    );
});

exports.default = AppMenu;
