// d2-ui/lib/app/AppBar.component.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('../header-bar/translate');

require('../header-bar/menu');

require('../header-bar/menu-ui');

var _headerBarDhis2 = require('../header-bar/dhis2');

var _headerBarDhis22 = _interopRequireDefault(_headerBarDhis2);

function stripApi(baseUrl) {
    return baseUrl.replace(/\/api\/?$/, '');
}

exports['default'] = _react2['default'].createClass({
    displayName: 'AppBar.component',

    contextTypes: {
        d2: _react2['default'].PropTypes.object.isRequired
    },

    getInitialState: function getInitialState() {
        return {
            headerBar: {}
        };
    },

    componentDidMount: function componentDidMount() {
        _headerBarDhis22['default'].settings.baseUrl = stripApi(this.context.d2.Api.getApi().baseUrl);
        _headerBarDhis22['default'].menu.ui.initMenu();
    },

    render: function render() {
        var headerBarStyle = {
            height: '44px',
            position: 'fixed',
            zIndex: 15,
            top: 0,
            left: 0,
            right: 0,
            boxShadow: '0 0 3px #222'
        };

        return _react2['default'].createElement(
            'div',
            { style: headerBarStyle, id: 'header' },
            _react2['default'].createElement(
                'a',
                { href: this.state.headerLink, title: this.state.headerBar.title, className: 'title-link' },
                _react2['default'].createElement('img', { className: 'header-logo', src: this.state.headerBar.logo, id: 'headerBanner' }),
                _react2['default'].createElement(
                    'span',
                    { className: 'header-text', id: 'headerText' },
                    this.state.headerBar.title
                )
            ),
            _react2['default'].createElement(
                'div',
                null,
                this.state.headerBar.message
            ),
            _react2['default'].createElement('div', { id: 'dhisDropDownMenu' })
        );
    }
});
module.exports = exports['default'];
