// d2-ui/lib/form-fields/theme.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _colors = require('material-ui/lib/styles/colors');

var _colors2 = _interopRequireDefault(_colors);

var _colorManipulator = require('material-ui/lib/utils/color-manipulator');

var _colorManipulator2 = _interopRequireDefault(_colorManipulator);

var _spacing = require('material-ui/lib/styles/spacing');

var _spacing2 = _interopRequireDefault(_spacing);

var _themeManager = require('material-ui/lib/styles/theme-manager');

var _themeManager2 = _interopRequireDefault(_themeManager);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var theme = {
    spacing: _spacing2.default,
    fontFamily: 'Roboto, sans-serif',
    palette: {
        primary1Color: _colors2.default.blue500,
        primary2Color: _colors2.default.blue700,
        primary3Color: _colors2.default.blue100,
        accent1Color: _colors2.default.orange500,
        accent2Color: _colors2.default.grey100,
        accent3Color: _colors2.default.grey500,
        textColor: _colors2.default.darkBlack,
        alternateTextColor: _colors2.default.white,
        canvasColor: _colors2.default.white,
        borderColor: _colors2.default.grey400,
        disabledColor: _colorManipulator2.default.fade(_colors2.default.darkBlack, 0.3)
    }
};

var muiTheme = _themeManager2.default.getMuiTheme(theme);

exports.default = Object.assign({}, muiTheme, {
    forms: {
        minWidth: 350,
        maxWidth: 750
    },
    formInput: {
        fontWeight: 100
    }
});
